front/components/agent_builder/instructions/AgentBuilderInstructionsEditor.tsx

```diff
@@ -0,0 +1,147 @@
+import { CharacterCount } from "@tiptap/extension-character-count";
+import Document from "@tiptap/extension-document";
+import { History } from "@tiptap/extension-history";
+import Text from "@tiptap/extension-text";
+import { EditorContent, useEditor } from "@tiptap/react";
+import { cva } from "class-variance-authority";
+import React, { useEffect, useMemo } from "react";
+
+import { useAgentBuilderInstructionsContext } from "@app/components/agent_builder/instructions/AgentBuilderInstructionsContext";
+import { ParagraphExtension } from "@app/components/assistant/conversation/input_bar/editor/extensions/ParagraphExtension";
+import {
+  plainTextFromTipTapContent,
+  tipTapContentFromPlainText,
+} from "@app/lib/client/assistant_builder/instructions";
+import { classNames } from "@app/lib/utils";
+
+export const INSTRUCTIONS_MAXIMUM_CHARACTER_COUNT = 120_000;
+
+const editorVariants = cva(
+  [
+    "overflow-auto min-h-60 h-full border rounded-xl p-2",
+    "transition-all duration-200",
+    "bg-muted-background dark:bg-muted-background-night",
+  ],
+  {
+    variants: {
+      error: {
+        true: [
+          "border-warning-500 dark:border-warning-500-night",
+          "focus:ring-warning-500 dark:focus:ring-warning-500-night",
+          "focus:outline-warning-500 dark:focus:outline-warning-500-night",
+          "focus:border-warning-500 dark:focus:border-warning-500-night",
+        ],
+        false: [
+          "border-border dark:border-border-night",
+          "focus:ring-highlight-300 dark:focus:ring-highlight-300-night",
+          "focus:outline-highlight-200 dark:focus:outline-highlight-200-night",
+          "focus:border-highlight-300 dark:focus:border-highlight-300-night",
+        ],
+      },
+    },
+    defaultVariants: {
+      error: false,
+    },
+  }
+);
+
+export function AgentBuilderInstructionsEditor() {
+  const { instructions, setInstructions } =
+    useAgentBuilderInstructionsContext();
+
+  const extensions = useMemo(
+    () => [
+      Document,
+      Text,
+      ParagraphExtension,
+      History,
+      CharacterCount.configure({
+        limit: INSTRUCTIONS_MAXIMUM_CHARACTER_COUNT,
+      }),
+    ],
+    []
+  );
```

Comment:
As there is no dependencies, I believe using `useRef` make more sense. Or even better outside of React, not sure it's necessary to be inside.